examples/http/client.cpp

```diff
@@ -0,0 +1,125 @@
+#include "caf/event_based_actor.hpp"
+#include "caf/ipv4_address.hpp"
+#include "caf/scheduled_actor/flow.hpp"
+#include "caf/span.hpp"
+#include "caf/uuid.hpp"
+
+#include <cassert>
+#include <iostream>
+#include <memory>
+#include <utility>
+
+using namespace std::literals;
+
+namespace http = caf::net::http;
+namespace ssl = caf::net::ssl;
+using namespace caf;
+
+// -- constants ----------------------------------------------------------------
+
+// static constexpr uint16_t default_port = 7788;
+// static constexpr std::string_view default_host = "localhost";
+
+constexpr auto default_method = http::method::get;
+
+// -- configuration setup ------------------------------------------------------
+struct config : actor_system_config {
+  config() {
+    opt_group{custom_options_, "global"}
+      .add<http::method>("method,m", "HTTP method to use")
+      .add<uri>("resource,r", "Requested resource (URL)")
+      .add<std::string>("payload,p", "Optional payload to send");
+    opt_group{custom_options_, "tls"} //
+      .add<bool>("enable", "enables encryption via TLS")
+      .add<std::string>("ca-file", "CA file for trusted servers");
+  }
+
+  caf::settings dump_content() const override {
+    auto result = actor_system_config::dump_content();
+    caf::put_missing(result, "method", default_method);
+    caf::put_missing(result, "payload", "");
+    return result;
+  }
+};
+
+auto defatul_port(std::string_view scheme) {
+  if (scheme == "http")
+    return 80;
+  else if (scheme == "https")
+    return 443;
+  else
+    return 0;
+}
+
+int caf_main(caf::actor_system& sys, const config& cfg) {
+  auto maybe_resource = caf::get_as<uri>(cfg, "resource");
+  if (!maybe_resource) {
+    std::cerr << "*** missing mandatory option --resource" << std::endl;
+    return EXIT_FAILURE;
+  }
+  auto resource = std::move(*maybe_resource);
+  auto use_ssl = caf::get_or(cfg, "tls.enable", false);
+  auto ca_file = caf::get_as<std::string>(cfg, "tls.ca-file");
+  if (resource.scheme() != "http" && resource.scheme() != "https") {
+    std::cerr << "*** only HTTP is supported at the moment" << std::endl;
+    return EXIT_FAILURE;
+  }
+  auto method = caf::get_or(cfg, "method", default_method);
+  auto payload = caf::get_or(cfg, "payload", ""sv);
+  auto& auth = resource.authority();
+  auto port = auth.port != 0 ? auth.port : defatul_port(resource.scheme());
+  auto conn
+    = http::with(sys)
+        // Optionally enable TLS.
+        .context(ssl::context::enable(use_ssl)
+                   .and_then(ssl::emplace_client(ssl::tls::v1_2))
+                   .and_then(ssl::load_verify_file_if(ca_file)))
+        // Connect to "$host:$port".
+        .connect(std::get<0>(auth.host), port)
+        // If we don't succeed at first, try up to 5 times with 1s delay.
+        .retry_delay(1s)
+        .max_retry_count(5)
+        // After connecting, send a get request.
+        .request(method, resource.path_query_fragment(), payload);
```

Comment:
Again, I'd just pass the URI along. The user shouldn't care about the nitty-gritty details. Ideally, the `request` should be overloaded to allow both a path (as string) and an URI.

```suggestion
        .request(method, resource, payload);
```

Reply:
That's something that was on my mind. One thing confusing me here is that an URI contains everything needed to make the request, so having a code like this:
```
auto result = http::with(sys)
        .connect(resource)
        .get(resource);
```
I assume you suggest  `connect` only reading the address and port part of the resource, and `get` reading the path and query part? This would be easier to implement. 
Whats probably more intuitive is `connect` reading everything, like:
```
auto result = http::with(sys)
        .connect(resource)
        .get();
```
However this would require us overloading the DSL and splitting the API to different types when we connect to an URI vs when we connect to an host/port part.

Reply:
Hm, this brings up a good point. The URI doesn't just configure host and port. It also tells whether to use TLS or not. You're currently using the `context` factories for that, but that shouldn't be necessary either.

```c++
auto result = http::with(sys)
        .connect(resource)
        .get();
```

I really like that. That would probably cover >99% of use cases anyways. I'd go as far as to *only* implement the URI interface. You can always just assemble an URI, so I don't see much point in making the implementation complicated if the user can just put together an URI for `connect`.